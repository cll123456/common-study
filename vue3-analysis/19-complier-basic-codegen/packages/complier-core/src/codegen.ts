import { isString } from "shared"
import { NodeTypes } from "./ast"
import { CREATE_ELEMENT_BLOCK, helperNameMap, OPEN_BLOCK, TO_DISPLAY_STRING } from "./runtimeHelpers"

export function codegen(root) {
  const context = createCodegenNode(root)
  const { push, newLine } = context
  // import { toDisplayString as _toDisplayString } from "vue"
  genFunctionPreamble(context)
  // 放入 export function 
  push('export function ')

  genFuncNameAndParams(context)
  newLine()
  // 放入 return和root的content
  push(`return `)
  genNode(root.createCodegenNode, context)
  newLine()
  push('}')
  return {
    code: context.code
  }
}
/**
 * 处理import导入的函数
 * @param context 
 */
function genFunctionPreamble(context) {
  const { push, newLine } = context
  const vueBinging = 'vue'
  const helperFuncs = context.root.helpers
  if (helperFuncs.length) {
    const helperFuncStr = helperFuncs.map(p => `${p} : _${p}`).join(', ')
    push(`import { ${helperFuncStr} } from "${vueBinging}"`)
    newLine()
  }
}

/**
 * 处理节点
 * @param root 
 * @param context 
 */
function genNode(node, context) {
  const type = node.type
  switch (type) {
    case NodeTypes.TEXT:
      genTextNode(node, context)
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context)
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context)
      break
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context)
  }

}

/**
 * 创建上下文
 * @param root 
 * @returns 
 */
function createCodegenNode(root: any) {
  const context = {
    code: '',
    root: root,
    push(code: string) {
      context.code += code
    },
    newLine() {
      context.code += '\n'
    },
    helper(key) {
      return `_${helperNameMap[key]}`
    }
  }
  return context
}

/**
 * 生成函数名和参数
 * @param context 
 */
function genFuncNameAndParams(context: { code: string; root: any; push(code: string): void; newLine(): void }) {
  const { push } = context
  const paramArr = ['_ctx', '_cache']
  const paramStr = paramArr.join(', ')
  const functionName = 'render'
  push(`${functionName} (${paramStr}) {`)
}

function genTextNode(node, context: any) {
  const { push } = context
  push(`"${node.content}"`)
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context;
  const toDisplayStringName = helper(TO_DISPLAY_STRING)
  push(`${toDisplayStringName}(`)
  genNode(node.content, context)
  push(')')
}



function genExpression(node: any, context: any) {
  const { push } = context
  push(node.content)
}


function genElement(node, context) {
  const { push, helper } = context
  const { tag, props, children } = node
  const openBlockName = helper(OPEN_BLOCK)
  const createElementBlockName = helper(CREATE_ELEMENT_BLOCK)
  push(`(${openBlockName}(), ${createElementBlockName}(`)
  if (children.length) {
    const argsNode = genNullableArgs([tag, props, children])
    genNodeList(argsNode, context)
  } else {
    push(`${tag}`)
  }
  push(`))`)
}


function genNullableArgs(args) {
  let length = args.length
  // 找到局部为不为null，然后把不存在的直接变成null
  while (length--) {
    if (args[length] != null) break
  }

  return args.slice(0, length + 1).map(arg => arg || `null`)
}

function genNodeList(argsNode: any, context: any) {
  const { push } = context
  for (let i = 0; i < argsNode.length; i++) {
    const node = argsNode[i]
    if (isString(node)) {
      push(`${node}`)
    } else if (Array.isArray(node)) {
      genNodeList(node, context)
    } else {
      genNode(node, context)
    }

    // 加上， 
    if (i < argsNode.length - 1) {
      push(', ')
    }
  }
}

function genCompoundExpression(node: any, context: any) {
  const { push } = context
  const { children } = node
  for (let i = 0; i < children.length; i++) {
    const child = children[i]
    if (isString(child)) {
      push(child)
    } else {
      genNode(child, context)
    }
  }
}
