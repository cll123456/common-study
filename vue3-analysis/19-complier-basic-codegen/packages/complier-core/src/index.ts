export { baseParse } from './parse'
export { codegen } from './codegen'
