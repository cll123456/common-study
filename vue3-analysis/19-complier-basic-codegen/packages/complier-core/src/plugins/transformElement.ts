import { NodeTypes } from "../ast"

export const transformElement = (node) => {
  return () => {
    if (node.type === NodeTypes.ELEMENT) {
      node.tag = `"${node.tag}"`
    }
  }
}
