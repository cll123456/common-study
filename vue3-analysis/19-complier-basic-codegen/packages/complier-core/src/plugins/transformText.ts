import { NodeTypes } from "../ast"

export const transformText = (node) => {
  return () => {
    if (node.type === NodeTypes.ELEMENT) {
      const children = node.children
      let currentContainer
      for (let i = 0; i < children.length; i++) {
        const child = children[i]
        if (isText(child)) {
          // 判断下一个节点是不是text
          for (let j = i + 1; j < children.length; j++) {
            const next = children[j]
            if (isText(next)) {
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [child]
                }
              }
              // 放入 +  放入next
              currentContainer.children.push(' + ', next);
  
              // 删除当前节点j
              children.splice(j, 1)
              j--
  
            } else {
              currentContainer = undefined
              break;
            }
          }
        }
      }
    }
  }
}



function isText(node) {
  return node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION
}
