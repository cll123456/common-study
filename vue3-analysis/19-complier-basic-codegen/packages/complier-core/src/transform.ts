import { NodeTypes } from "./ast"
import { CREATE_ELEMENT_BLOCK, helperNameMap, OPEN_BLOCK, TO_DISPLAY_STRING } from "./runtimeHelpers"

export function transform(root, options = {}) {
  // 创建当前transform的上下文
  const context = createTransformContext(root, options)
  // 树的深度优先遍历
  traverseNode(root, context)
  // 生成node
  createRootCodegen(context.root)

  root.helpers.push(...context.helpers.keys())
}
/**
 * 树的深度优先遍历
 * @param context 
 */
function traverseNode(node, context) {
  // 执行插件
  const { nodeTransforms } = context
  // 当执行退出的时候来执行插件
  const exitFn = []
  if (nodeTransforms && nodeTransforms.length > 0) {
    // 遍历nodeTransforms
    for (let j = 0; j < nodeTransforms.length; j++) {
      const nodeTransform = nodeTransforms[j]
      // 调用nodeTransform
     const fn = nodeTransform(node)
     if(fn){
        exitFn.push(fn)
     }
    }
  }

  const type = node.type
  switch (type) {
    case NodeTypes.INTERPOLATION:
      // 把import { toDisplayString } from "vue"放入
      context.helper(helperNameMap[TO_DISPLAY_STRING])
      break;
    case NodeTypes.ROOT:
      traverseChildren(node.children, context)
      break;
    case NodeTypes.ELEMENT:
      context.helper(helperNameMap[CREATE_ELEMENT_BLOCK])
      context.helper(helperNameMap[OPEN_BLOCK])
      traverseChildren(node.children, context)
      break;
    default:
  }
  // 执行收集的插件
  let i = exitFn.length
  while (i--) {
    exitFn[i]()
  }

}

function traverseChildren(children, context) {
  // 存在children,优先遍历
  if (children) {
    for (let i = 0; i < children.length; i++) {
      traverseNode(children[i], context)
    }
  }
}




/**
 * 创建transform的上下文
 * @param root 
 */
function createTransformContext(root: any, options) {
  root.helpers = []
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, true)
    }
  }

  return context
}

/**
 * 创建节点生成
 * @param root 
 */
function createRootCodegen(root: any) {
  root.createCodegenNode = root.children[0]
}

