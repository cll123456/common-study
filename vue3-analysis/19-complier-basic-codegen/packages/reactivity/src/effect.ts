import { extend } from 'shared'


// 当前执行的effect
let activeEffect = void 0;
// 是否需要进行依赖收集
let shouldTrack = false;

export class EffectReactive {
  fn: Function;
  // 保存正则执行的effect,用于清除
  deps = []
  // 防止在同一时间调用多次stop
  active = true;
  // stop 回调函数
  public onStop: Function;

  constructor(fn, public scheduler?) {
    this.fn = fn
    this.scheduler = scheduler
  }

  run() {
    // 调用stop后不需要收集依赖
    if (!this.active) {
      activeEffect = this;
      return this.fn()
    }

    // 收集依赖
    shouldTrack = true;
    activeEffect = this;

    const result = this.fn();

    // 执行fn后停止收集依赖
    shouldTrack = false;

    return result

  }

  /**
   * 停止响应
   */
  stop() {
    if (this.active) {

      cleanEffect(this);
      if (this.onStop) {
        this.onStop()
      }
      /**
       * 防止在同一个activeEffect 中重复调用stop
       */
      this.active = false
    }
  }
}
/**
 * 清空当前收集的effect
 * @param effect 
 */
function cleanEffect(effect) {
  effect.deps.forEach(effectSet => {
    effectSet.delete(effect)
  })
  effect.deps.length = 0
}

/**
 * 创建一个effect
 * @param fn 
 */
export function effect(fn, options: any = {}) {
  const _effect = new EffectReactive(fn, options.scheduler)
  // 可能还有其他参数需要合并
  // _effect.onStop = options.onStop
  extend(_effect, options)
  _effect.run()
  const runner = _effect.run.bind(_effect)
  runner.effect = _effect
  return runner
}


const targetMap = new WeakMap();

/**
 * 收集依赖
 * @param target 
 * @param key 
 */
export function track(target, key) {
  if (!tracking()) return

  let depsMap = targetMap.get(target);
  if (!depsMap) {
    depsMap = new Map();
    targetMap.set(target, depsMap);
  }
  // 获取key
  let deps = depsMap.get(key);
  if (!deps) {
    deps = new Set();
    depsMap.set(key, deps);
  }
  trackEffect(deps)
}

/**
 * 收集effect
 * @param deps 
 * @returns 
 */
export function trackEffect(deps) {
  // 存在的话，不需要反复收集
  if (deps.has(activeEffect)) return
  // 收集依赖
  deps.add(activeEffect)
  activeEffect.deps.push(deps)
}

/**
 * 是否正在进行依赖收集
 * @returns 
 */
export function tracking() {
  return shouldTrack && activeEffect
}

/**
 * 依赖触发
 * @param target 
 * @param key 
 * @returns 
 */
export function trigger(target, key) {
  let depsMap = targetMap.get(target)
  if (!depsMap) {
    return
  }
  let deps = depsMap.get(key);
  if (!deps) {
    return
  }
  // 触发依赖
  triggerEffect(deps)
}
/**
 * 遍历触发依赖
 * @param deps 
 */
export function triggerEffect(deps) {
  deps.forEach(effect => {
    if (effect.scheduler) {
      effect.scheduler()
    } else {
      effect.run()
    }
  })
}

/**
 * 停止响应式更新
 * @param runner 
 */
export function stop(runner) {
  runner.effect.stop()
}


