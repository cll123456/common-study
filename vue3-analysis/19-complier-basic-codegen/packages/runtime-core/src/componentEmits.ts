import { camize, capitalize, handlerName } from "shared"

export function emit(instance, event, ...args) {
  const { props } = instance
  // 判断props里面是否有对应的事件，有的话执行，没有就不执行
  const key = handlerName(capitalize(camize(event)))
  const handler = props[key]

  handler && handler(...args)
}
