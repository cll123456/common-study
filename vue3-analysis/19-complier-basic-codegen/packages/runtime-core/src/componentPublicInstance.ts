import { isOwn } from "shared"

const publicPropertiesMap = {
  // 返回当前组件的el
  $el: (instance) => instance.vnode.el,
  // 返回当前组件的slots
  $slots: (instance) => instance.slots,
  // 返回当前组件的props
  $props: (instance) => instance.props
}

export const publicInstanceHandles = {
  get({ _: instance }, key) {
    // 判断key是否在组件的setupState上
    if (isOwn(instance.setupState, key)) {
      return instance.setupState[key]
    } else if (isOwn(instance.props, key)) {
      // 在组件的属性上
      return instance.props[key]
    }
    if (key in publicPropertiesMap) {
      return publicPropertiesMap[key](instance)
    }
  }
}
