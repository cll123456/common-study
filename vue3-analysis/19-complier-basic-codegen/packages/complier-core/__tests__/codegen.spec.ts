import { baseParse, codegen } from "../src"
import { transform } from "../src/transform"
import { transformInterpolation } from '../src/plugins/transformInterpolation'
import { transformElement } from '../src/plugins/transformElement'
import { transformText } from '../src/plugins/transformText'

describe('test codegen', () => {

  test('code gen str ---> hi', () => {
    const ast = baseParse('hi')
    transform(ast)
    const codeObj = codegen(ast)

    expect(codeObj.code).toMatchSnapshot()
  })

  test('code gen interpolation  ---> {{message}}', () => {
    const ast = baseParse('{{message}}')
    transform(ast, {
      nodeTransforms: [transformInterpolation]
    })
    const codeObj = codegen(ast)
    expect(codeObj.code).toMatchSnapshot()
  })

  test('code gen element  ---> <div></div>', () => {
    const ast = baseParse('<div></div>')
    transform(ast, {
      nodeTransforms: [transformElement]
    })
    const codeObj = codegen(ast)
    expect(codeObj.code).toMatchSnapshot()
  })

  /**
     * const { openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue

return function render(_ctx, _cache, $props, $setup, $data, $options) {
  return (_openBlock(), _createElementBlock("div", null, "hi twinkle,"))
}
     */
  test('code gen element and  str  ---> <div>hi twinkle,</div>', () => {
    const ast = baseParse('<div>hi, twinkle,</div>')
    transform(ast, {
      nodeTransforms: [transformElement]
    })
    const codeObj = codegen(ast)
    expect(codeObj.code).toMatchSnapshot()
  })


  /**
   * 
   */
  test('code gen element and  interpolation  ---> <div>{{message}}</div>', () => {
    const ast = baseParse('<div>{{message}}</div>')
    transform(ast, {
      nodeTransforms: [transformElement, transformInterpolation]
    })
    const codeObj = codegen(ast)
    expect(codeObj.code).toMatchSnapshot()
  })


  test('code gen element, str and  interpolation  ---> <div>hi twinkle, {{message}}</div>', () => {
    const ast = baseParse('<div>hi, twinkle, {{message}}</div>')

    transform(ast, {
      nodeTransforms: [transformText, transformElement, transformInterpolation]
    })
    const codeObj = codegen(ast)
    expect(codeObj.code).toMatchSnapshot()
  })
})
