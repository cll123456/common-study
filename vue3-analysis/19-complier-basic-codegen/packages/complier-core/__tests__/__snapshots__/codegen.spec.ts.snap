// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test codegen code gen element  ---> <div></div> 1`] = `
"import { createElementBlock : _createElementBlock, openBlock : _openBlock } from \\"vue\\"
export function render (_ctx, _cache) {
return (_openBlock(), _createElementBlock(\\"div\\"))
}"
`;

exports[`test codegen code gen element and  interpolation  ---> <div>{{message}}</div> 1`] = `
"import { createElementBlock : _createElementBlock, openBlock : _openBlock, toDisplayString : _toDisplayString } from \\"vue\\"
export function render (_ctx, _cache) {
return (_openBlock(), _createElementBlock(\\"div\\", null, _toDisplayString(_ctx.message)))
}"
`;

exports[`test codegen code gen element and  str  ---> <div>hi twinkle,</div> 1`] = `
"import { createElementBlock : _createElementBlock, openBlock : _openBlock } from \\"vue\\"
export function render (_ctx, _cache) {
return (_openBlock(), _createElementBlock(\\"div\\", null, \\"hi, twinkle,\\"))
}"
`;

exports[`test codegen code gen element, str and  interpolation  ---> <div>hi twinkle, {{message}}</div> 1`] = `
"import { createElementBlock : _createElementBlock, openBlock : _openBlock, toDisplayString : _toDisplayString } from \\"vue\\"
export function render (_ctx, _cache) {
return (_openBlock(), _createElementBlock(\\"div\\", null, \\"hi, twinkle, \\" + _toDisplayString(_ctx.message)))
}"
`;

exports[`test codegen code gen interpolation  ---> {{message}} 1`] = `
"import { toDisplayString : _toDisplayString } from \\"vue\\"
export function render (_ctx, _cache) {
return _toDisplayString(_ctx.message)
}"
`;

exports[`test codegen code gen str ---> hi 1`] = `
"export function render (_ctx, _cache) {
return \\"hi\\"
}"
`;
